components/script/dom/request.rs

```diff
@@ -308,21 +308,26 @@ impl Request {
         // Step 27
         let mut headers_copy = r.Headers();
 
-        // This is equivalent to the specification's concept of
-        // "associated headers list".
-        if let RequestInfo::Request(ref input_request) = input {
-            headers_copy = input_request.Headers();
-        }
-
         // Step 28
         if let Some(possible_header) = init.headers.as_ref() {
-            if let &HeadersInit::Headers(ref init_headers) = possible_header {
-                headers_copy = init_headers.clone();
+            match possible_header {
+                &HeadersInit::Headers(ref init_headers) => {
+                    headers_copy = init_headers.clone();
+                }
+                &HeadersInit::ByteStringSequenceSequence(ref init_sequence) => {
+                    try!(headers_copy.fill(Some(
+                        HeadersInit::ByteStringSequenceSequence(init_sequence.clone()))));
+                },
+                &HeadersInit::ByteStringMozMap(ref init_map) => {
+                    try!(headers_copy.fill(Some(
+                        HeadersInit::ByteStringMozMap(init_map.clone()))));
+                },
             }
         }
 
         // Step 29
-        r.Headers().empty_header_list();
+        // When r.Headers()'s header list is emptied,
+        // r.Headers that was filled in Step 28 becomes empty.
```

Comment:
Hmm. I realized I wasn't making a deep copy. I had to do `r.Headers.r().clone()` in order to actually make a deep copy.

With this approach though, because `r.Headers()` is a `Root<T>`, it becomes difficult to make a mutable reference to `r.Headers()`. Without the mutable reference, we cannot set `r.Headers()` to be the deep copied headers in Step 27.

I want to suggest we leave the comment more descriptive (as you wrote above), but not make a deep copy in Step 27.